### ISSUE TITLE
[ios][expo-av] Crash - EXC_BAD_ACCESS

### ISSUE THREAD

**Issue by hirbod** [expo/expo#16982]
### Summary

Sentry reporting

```
EXC_BAD_ACCESS: isEqual: > progressUpdateIntervalMillis >
Attempted to dereference garbage pointer 0x37bfdd1010408.
```

<img width="954" alt="Bildschirmfoto 2022-04-09 um 13 24 38" src="https://user-images.githubusercontent.com/504909/162569801-11de2c62-9c98-42c0-9ac7-ae8be0592a10.png">

<img width="507" alt="Bildschirmfoto 2022-04-09 um 13 28 37" src="https://user-images.githubusercontent.com/504909/162569887-842445dc-1d6b-408e-b5f3-0744cffdd89f.png">

I am sorry if this isn't helpful, but that is all I have from Sentry.

I am using expo-av 11.1.0 (custom-dev-client)

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

bare

### What platform(s) does this occur on?

iOS

### SDK Version (managed workflow only)

44

### Environment

```
  expo-env-info 1.0.2 environment info:
    System:
      OS: macOS 12.2.1
      Shell: 5.8 - /bin/zsh
    Binaries:
      Node: 14.19.1 - /usr/local/opt/node@14/bin/node
      Yarn: 1.22.17 - /usr/local/bin/yarn
      npm: 6.14.16 - /usr/local/opt/node@14/bin/npm
    Managers:
      CocoaPods: 1.11.2 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 21.4, iOS 15.4, macOS 12.3, tvOS 15.4, watchOS 8.5
    IDEs:
      Android Studio: 4.1 AI-201.8743.12.41.6953283
      Xcode: 13.3/13E113 - /usr/bin/xcodebuild
    npmGlobalPackages:
      eas-cli: 0.41.1
      expo-cli: 5.3.0
    Expo Workflow: bare
```

### Reproducible demo

I have some crash reports in Sentry. I've never encountered them myself, thus I can't provide a reproducible example.

**Comment by Pickleboyonline** [comment]
We have received a memory issue similar to yours but with the `_statusToSet` object in `EXVideoView.m`. What happens is that the Imperative Playback API and props mutate _statusToSet on the different threads and causes a race condition. It's a bit difficult to reproduce, but let's say if you repetitively pause/play an Expo Video using props and use its ref to seek at the same you can get a crash like this.

The specific error is an `SIGSEGV: Attempted to dereference garbage pointer 0xe190a19d9740.` that happens here:
```
-[EXVideoView _tryUpdateDataStatus:rejecter:]
EXVideoView.m, line 96
```
On that line the message `[_statusToSet removeAllObjects];` is run and the Imperative Playback API and prop updates are running this message at the same time, with the Imperative Playback API running on the Main thread and props updates being dispatched to the `_exAV.methodQueue` serial queue.

As a solution, I forced the imperative Playback API to dispatch [EXVideoView setStatus:resolver:rejected:]; on the `_exAV.methodQueue` serial queue more thread-safe. So far we have yet to see this crash again after doing so. But now that I look at this again, why not ensure that updating the status through props runs on the main thread instead? Why are props events dispatched to `_exAV.methodQueue` serial queue to begin with? Or, maybe use a thread-safe implementation of NSMutableDictionary (an example [here](https://www.guiguan.net/ggmutabledictionary-thread-safe-nsmutabledictionary/))

Additionally, I feel that auditing the thread safety of updating the EXVideoView should be examined for use cases outside of updating the status, such as setting the source. It appears that your memory-related issue revolves around setting the source. Perhaps `parameters` NSMutableDictionary is being mutated on different threads? 

Finally, we recently received another error most likely due to a race condition were the EXAVPlayerData is only removed as an observer on the `dealloc` method, which can run on a background thread while a message is being sent to the object on the main thread through KVO. Here is an example stack trace:
```
NSInternalInconsistencyException: <EXAVPlayerData: 0x2800f3400>: An -observeValueForKeyPath:ofObject:change:context: message was received but not handled.
Key path: status
Observed object: <AVQueuePlayer: 0x283b96c40>
Change: {
    kind = 1;
}
Context: 0x0

0  CoreFoundation    ___exceptionPreprocess
1  libobjc.A.dylib   _objc_exception_throw
2  Foundation        -[NSObject(NSKeyValueObserving) observeValueForKeyPath:ofObject:change:context:]
3  App      -[EXAVPlayerData observeValueForKeyPath:ofObject:change:context:] (EXAVPlayerData.m:694:5)
4  Foundation        _NSKeyValueNotifyObserver
5  Foundation        _NSKeyValueDidChange
6  Foundation        _NSKeyValueDidChangeWithPerThreadPendingNotifications
7  AVFCore           ___67-[AVPlayer _createAndConfigureFigPlayerWithType:completionHandler:]_block_invoke_2
8  libdispatch.dylib __dispatch_call_block_and_release
9  libdispatch.dylib __dispatch_client_callout
10 libdispatch.dylib __dispatch_main_queue_callback_4CF
11 CoreFoundation    ___CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__
12 CoreFoundation    ___CFRunLoopRun
13 CoreFoundation    _CFRunLoopRunSpecific
14 GraphicsServices  _GSEventRunModal
15 UIKitCore         -[UIApplication _run]
16 UIKitCore         _UIApplicationMain
17 App      main (main.m:7:12)
18 dyld              start
```
How I chose to fix this was to create a `-(void)cleanup` function on the EXAVPlayerData object that had all the code in the `-(void)dealloc` block. I made the `dealloc` block call the `cleanup` method instead. In `EXVideoView.m`, I explicitly call `[_data cleanup];` inside of `[EXVideoView _removeData:];` before setting `_data` equal to nil. I then wrapped the code in `[EXVideoView _removeData:];` inside of a block that runs synchronously on the main-thread, similar to `[EXVideoView _removePlayer:];`. This change should be able to remove EXAVPlayerData as an observer before a message is sent to it on the main thread. 

Here's a look at everything I modified so far in the expo-av module to address these issues (ignore the line where I comment out `[session setActive:NO error:&error];` in `EXAudioSessionManager.m` as it looks like this got patched in expo-av version 11.1.0):

```patch
diff --git a/node_modules/expo-av/ios/EXAV/EXAV.m b/node_modules/expo-av/ios/EXAV/EXAV.m
index edf934f..902e6c4 100644
--- a/node_modules/expo-av/ios/EXAV/EXAV.m
+++ b/node_modules/expo-av/ios/EXAV/EXAV.m
@@ -808,7 +808,7 @@ - (void)sendEventWithName:(NSString *)eventName body:(NSDictionary *)body
                     rejecter:(EXPromiseRejectBlock)reject)
 {
   [self _runBlock:^(EXVideoView *view) {
-    [view setStatus:status resolver:resolve rejecter:reject];
+    [view setStatusFromPlayBackAPI:status resolver:resolve rejecter:reject];
   } withEXVideoViewForTag:reactTag withRejecter:reject];
 }
 
diff --git a/node_modules/expo-av/ios/EXAV/EXAVPlayerData.h b/node_modules/expo-av/ios/EXAV/EXAVPlayerData.h
index 1a9a68f..96049fe 100644
--- a/node_modules/expo-av/ios/EXAV/EXAVPlayerData.h
+++ b/node_modules/expo-av/ios/EXAV/EXAVPlayerData.h
@@ -32,4 +32,6 @@
                 resolver:(EXPromiseResolveBlock)resolve
                 rejecter:(EXPromiseRejectBlock)reject;
 
+-(void)cleanup;
+
 @end
diff --git a/node_modules/expo-av/ios/EXAV/EXAVPlayerData.m b/node_modules/expo-av/ios/EXAV/EXAVPlayerData.m
index 7c83d85..c5bca3e 100644
--- a/node_modules/expo-av/ios/EXAV/EXAVPlayerData.m
+++ b/node_modules/expo-av/ios/EXAV/EXAVPlayerData.m
@@ -1066,12 +1066,17 @@ - (void)handleMediaServicesReset:(void (^)(void))finishCallback
 
 #pragma mark - NSObject Lifecycle
 
+-(void)cleanup
+{
+    // this triggers the audio tap removal
+    [self setSampleBufferCallback:nil];
+    [self _removeTimeObserver];
+    [self _removeObservers];
+}
+
 - (void)dealloc
 {
-  // this triggers the audio tap removal
-  [self setSampleBufferCallback:nil];
-  [self _removeTimeObserver];
-  [self _removeObservers];
+    [self cleanup];
 }
 
 # pragma mark - Utilities
diff --git a/node_modules/expo-av/ios/EXAV/EXAudioSessionManager.m b/node_modules/expo-av/ios/EXAV/EXAudioSessionManager.m
index edd11ed..24acc89 100644
--- a/node_modules/expo-av/ios/EXAV/EXAudioSessionManager.m
+++ b/node_modules/expo-av/ios/EXAV/EXAudioSessionManager.m
@@ -175,7 +175,21 @@ - (NSError *)_updateSessionConfiguration
   // And if the session should be activated, let's configure it first!
 
   if (!shouldBeActive && _sessionIsActive) {
-    [session setActive:NO error:&error];
+    
+   /*
+    *   Let's not deactivate the session for now and rely on mixing to achieve
+    *   the default behavior of NO MIXIN and MIXIN for when the video player is
+    *   playing and not playing, respectively. The reason for this approach is that
+    *   calling [ :pause] on a player does not immediately pause the player, and calling
+    *   [session setActive:NO error:&error]; will cause an error because the video
+    *   will still be playing within milliseconds of pausing. The error will also
+    *   block the UI thread as a result.
+    *
+    *   In general this practice is OK and actually recommended by the official docs.
+    *   The only real reason to disable the audio session is if the app plays sounds
+    *   in the background, which is not our use case.
+    */
+    // [session setActive:NO error:&error];
     if (!error) {
       _sessionIsActive = NO;
     }
diff --git a/node_modules/expo-av/ios/EXAV/Video/EXVideoView.h b/node_modules/expo-av/ios/EXAV/Video/EXVideoView.h
index f4e5f0e..7cbe9af 100644
--- a/node_modules/expo-av/ios/EXAV/Video/EXVideoView.h
+++ b/node_modules/expo-av/ios/EXAV/Video/EXVideoView.h
@@ -32,6 +32,10 @@ typedef NS_OPTIONS(NSUInteger, EXVideoFullscreenUpdate)
          resolver:(EXPromiseResolveBlock)resolve
          rejecter:(EXPromiseRejectBlock)reject;
 
+- (void)setStatusFromPlayBackAPI:(NSDictionary *)status
+           resolver:(EXPromiseResolveBlock)resolve
+           rejecter:(EXPromiseRejectBlock)reject;
+
 - (void)replayWithStatus:(NSDictionary *)status
                 resolver:(EXPromiseResolveBlock)resolve
                 rejecter:(EXPromiseRejectBlock)reject;
diff --git a/node_modules/expo-av/ios/EXAV/Video/EXVideoView.m b/node_modules/expo-av/ios/EXAV/Video/EXVideoView.m
index 488677a..26de8fc 100644
--- a/node_modules/expo-av/ios/EXAV/Video/EXVideoView.m
+++ b/node_modules/expo-av/ios/EXAV/Video/EXVideoView.m
@@ -120,12 +120,23 @@ - (void)_updateForNewPlayer
 
 - (void)_removeData
 {
-  if (_data) {
-    [_data pauseImmediately];
-    [_data setStatusUpdateCallback:nil];
-    [_exAV demoteAudioSessionIfPossible];
-    _data = nil;
-  }
+    EX_WEAKIFY(self);
+    void (^block)(void) = ^ {
+        EX_ENSURE_STRONGIFY(self)
+        if (self->_data) {
+            // We must remove _data as an observer synchronously on the main thread
+            // to prevent race conditions.
+            // dealloc tends to run on a background thread so we need to explicitly
+            // run _data.cleanup on the main thread
+            [self->_data cleanup];
+            [self->_data pauseImmediately];
+            [self->_data setStatusUpdateCallback:nil];
+            [self->_exAV demoteAudioSessionIfPossible];
+            self->_data = nil;
+        }
+    };
+    
+    [EXUtilities performSynchronouslyOnMainThread:block];
 }
 
 - (void)_removePlayer
@@ -561,6 +572,17 @@ - (void)setStatus:(NSDictionary *)status
   });
 }
 
+- (void)setStatusFromPlayBackAPI:(NSDictionary *)status
+           resolver:(EXPromiseResolveBlock)resolve
+           rejecter:(EXPromiseRejectBlock)reject;
+{
+  EX_WEAKIFY(self);
+  dispatch_async(_exAV.methodQueue, ^{
+    EX_ENSURE_STRONGIFY(self);
+    [self setStatus:status resolver:resolve rejecter:reject];
+  });
+}
+
 - (NSDictionary *)getStatus
 {
   if (_data) {
```




@joebernard

**Comment by hirbod** [comment]
@Pickleboyonline thank you very much for your efforts and explanations. I think you should provide a PR.

Let's pull in @mnightingale @bbarthec for their opinion.

**Comment by hirbod** [comment]
P.S: We already removed `[session setActive:NO error:&error];` in expo-av 11.1.0 as a result of plenty bug reports (sync blocking task)

**Comment by Pickleboyonline** [comment]
No problem, thank you for the response. I made and referenced the PR as requested.

**Comment by hirbod** [comment]
I see many of those crashes now (I build something like a TikTok / Instagram Reel) so  I am constantly creating and destroying expo av (max 3 video instances at a time). 

And if the app was backgrounded for too long, it will crash with that error.

**Comment by hirbod** [comment]
This was most likely fixed with #17036
